simpeg/utils/model_utils.py

```diff
@@ -150,3 +169,174 @@ def depth_weighting(
         wz = wz[active_cells]
 
     return wz / np.nanmax(wz)
+
+
+@njit(parallel=True)
+def _distance_weighting_numba(
+    cell_centers: np.ndarray,
+    reference_locs: np.ndarray,
+    threshold: float,
+    exponent: float = 2.0,
+) -> np.ndarray:
+    r"""
+    distance weighting kernel in numba.
+
+    If numba is not installed, this will work as a regular for loop.
+
+    Parameters
+    ----------
+    cell_centers : np.ndarray
+        cell centers of the mesh.
+    reference_locs : (n, ndim) numpy.ndarray
+        The coordinate of the reference location, usually the receiver locations,
+        for the distance weighting.
+        A 2d array, with multiple reference locations, where each row should
+        contain the coordinates of a single location point in the following
+        order: _x_, _y_, _z_ (for 3D meshes) or _x_, _z_ (for 2D meshes).
+    threshold : float
+        Threshold parameters used in the distance weighting.
+    exponent : float, optional
+        Exponent parameter for distance weighting.
+        The exponent should match the natural decay power of the potential
+        field. For example, for gravity acceleration, set it to 2; for magnetic
+        fields, to 3.
+
+    Returns
+    -------
+    (n_active) numpy.ndarray
+        Normalized distance weights for the mesh at every active cell as
+        a 1d-array.
+    """
+    n_active_cells = cell_centers.shape[0]
+    n_reference_locs = len(reference_locs)
+
+    distance_weights = np.zeros(n_active_cells)
+    for j in prange(n_active_cells):
+        cell_center = cell_centers[j]
+        for i in range(n_reference_locs):
+            reference_loc = reference_locs[i]
+            distance = np.sqrt(((cell_center - reference_loc) ** 2).sum())
+            distance_weights[j] += (distance + threshold) ** (-2 * exponent)
+
+    distance_weights = np.sqrt(distance_weights)
+    distance_weights /= np.nanmax(distance_weights)
+    return distance_weights
+
+
+def distance_weighting(
+    mesh: discretize.base.BaseMesh,
+    reference_locs: np.ndarray,
+    active_cells: Optional[np.ndarray] = None,
+    exponent: Optional[float] = 2.0,
+    threshold: Optional[float] = None,
+    engine: Literal["numba", "scipy"] = "numba",
```

Comment:
Usually not good to have a string be the default parameter,

I'd recommend:
* Default to None:
    * If `numba` is available, use that
    * Else use `scipy`

Reply:
I see your point. I think I'd prefer having code that run slow than code that might fill all available memory if the 2d array that scipy creates is too large.

I'd like to leave the default behavior as it is for now at least.

Reply:
If that’s so, the default should be a for loop over data points, with vectorized operations along the model cells. Not as fast as numba, but still memory efficient.

Reply:
But, future issues.

Reply:
You are right. Let's work on that later on, so we don't block this PR now.